lib/turn_endpoint.ex

```diff
@@ -0,0 +1,624 @@
+      if medium == max_port,
+        do: {medium, max_port},
+        else: {medium + 1, max_port}
+
+    [:udp, :tcp]
+    |> Enum.map(fn transport ->
+      secret = Utils.generate_secret()
+
+      {:ok, port, pid} =
+        Utils.start_integrated_turn(
+          secret,
+          client_port_range: client_port_range,
+          alloc_port_range: alloc_port_range,
+          ip: ip,
+          mock_ip: mock_ip,
+          transport: transport,
+          parent: connector,
+          fake_candidate_addr: @fake_candidate_addr,
+          elixir_ice_impl: true
+        )
+
+      %{
+        relay_type: transport,
+        secret: secret,
+        server_addr: ip,
+        mocked_server_addr: mock_ip,
+        server_port: port,
+        pid: pid
+      }
+    end)
+  end
+
+  defp start_integrated_turn_servers(_options, _connector), do: []
+
+  defp do_handle_connectivity_check(%{class: :request} = attrs, alloc_pid, ctx, state) do
+    alloc = state.turn_allocs[alloc_pid]
+
+    Utils.send_binding_success(
+      alloc_pid,
+      state.local_ice_pwd,
+      attrs.magic,
+      attrs.trid,
+      attrs.username
+    )
+
+    alloc = %{alloc | passed_check_from_browser: true, magic: attrs.magic}
+
+    if not alloc.passed_check_from_sfu do
+      trid = Utils.generate_transaction_id()
+      new_username = String.split(attrs.username, ":") |> Enum.reverse() |> Enum.join(":")
+
+      Utils.send_binding_request(
+        alloc_pid,
+        state.remote_ice_pwd,
+        attrs.magic,
+        trid,
+        new_username,
+        attrs.priority
+      )
+    end
+
+    alloc =
+      if attrs.use_candidate,
+        do: %{alloc | in_nominated_pair: true},
+        else: alloc
+
+    state = put_in(state, [:turn_allocs, alloc_pid], alloc)
+    maybe_select_allocation(alloc, ctx, state)
+  end
+
+  defp do_handle_connectivity_check(%{class: :response}, alloc_pid, ctx, state) do
+    alloc = state.turn_allocs[alloc_pid]
+    alloc = %{alloc | passed_check_from_sfu: true}
+    maybe_select_allocation(alloc, ctx, state)
+  end
+
+  defp do_handle_connectivity_check(%{class: :error}, _, _, state) do
+    {state, []}
+  end
+
+  defp maybe_select_allocation(
+         %{
+           passed_check_from_browser: true,
+           passed_check_from_sfu: true,
+           in_nominated_pair: true
+         } = alloc,
+         ctx,
+         state
+       ) do
+    if state.selected_alloc != alloc.pid do
+      select_alloc(alloc.pid, ctx, state)
+    else
+      {state, []}
+    end
+  end
+
+  defp maybe_select_allocation(_alloc, _ctx, state) do
+    {state, []}
+  end
+
+  defp select_alloc(alloc_pid, ctx, state) do
+    state = Map.put(state, :selected_alloc, alloc_pid)
+    Membrane.Logger.debug("Component #{@component_id} READY")
+
+    {hsk_state, hsk_status, _hsk_data} = state.handshake
+
+    if state.component_connected? and hsk_status != :finished do
+      {state, [notify: {:component_state_failed, @stream_id, @component_id}]}
```

Comment:
I don't get this `if` :thinking: If we select allocation then it means component is connected. Even if handshake is not finished yet we shouldn't send `component_state_failed`. I think we can completely remove this `if`.